Adds TLS support

tomponline opened the pull request:
Hi @repejota 

This is my initial commit to implement TLS support as request in https://github.com/repejota/phpnats/issues/121

Its pretty rough at the moment, e.g. its missing the ability to specify a custom CA cert, and does not verify the server's cert.

However I hope it is a basis for further development and an opportunity to check that you are happy with the approach.

Thanks
Tom

Commit:
Initial commit for adding TLS support

Commit:
Adds ability to use custom stream context to allow TLS mode

tomponline commented:
@repejota I've updated this PR to allow a custom stream context to be passed in, allowing any of the TLS options to be configured externally to the library. This has been added to the ConnectionOptions class. 

Example usage:

```php
<?php
require_once __DIR__.'/vendor/autoload.php';


$context = stream_context_create();
stream_context_set_option($context, 'ssl', 'verify_peer', true);
stream_context_set_option($context, 'ssl', 'cafile', '/var/lib/puppet/ssl/certs/ca.pem');

$connectionOptions = new \Nats\ConnectionOptions();
$connectionOptions->setHost('localhost')->setPort(4443)->setStreamContext($context);
$client = new \Nats\Connection($connectionOptions);
$client->connect();
// Simple Subscriber.
$client->subscribe(
    'foo',
    function ($message) {
        printf("Data: %s\r\n", $message->getBody());
    }
);

// Wait for 1 message.
$client->wait();
$client->close();
```

Commit:
whitespace

Commit:
Adds ability to use custom stream context to allow TLS mode

Commit:
adds default stream context

ancarda commented on src/Nats/Connection.php:
```diff
@@ -443,19 +444,33 @@ public function connect($timeout = null)
+            $this->processServerInfo($infoResponse);
+            if ($this->serverInfo->isTLSRequired()) {
+                set_error_handler(
+                    function ($errno, $errstr, $errfile, $errline) {
```
@tomponline: Should we include `restore_error_handler` here (before we throw)? Otherwise once this is caught, we won't have cleaned up the error handler we set, right?

tomponline replied:
added

ancarda commented:
So, I've been looking into why this pull request failed. What I suspect might be the cause is [a recent commit to NATS](https://github.com/nats-io/gnatsd/commit/17fecd4c9be6592de42d1ff5c8c2659a5d0382a5#diff-91bbeda7eb98a7adc57b9e47e2cf5c2b) (June 21st) adds `omitempty` to several fields including `auth_required`. This seems to cause undefined index notices when using NATS 1.2.0. This would normally not be an issue but the PHPUnit XML in this repo [converts notices to exceptions](https://github.com/repejota/phpnats/blob/develop/phpunit.xml.dist#L7).

This code was passing in master, possibly as the Docker image for NATS has been updated since the last commit 68e6d24232e8b5aaf134697e47de2652b3ebff6f (June 9th) which is before the aforementioned commit adding `omitempty`.

@repejota: Is there a way to -- and would you want to -- restrict which version of the Docker image you are using?

As I can't seem to commit directly here, I've prepared a patch file which @tomponline can apply to fix this.

Commit:
NATS 1.2.x support: catch not set fields

This commit improves support for NATS 1.2.x by checking if several
fields; `auth_required`, `tls_required`, and `tls_verify` are set in
the initial INFO packet. As of NATS 1.2, these fields aren't always
set, which can cause an E_NOTICE to be raised.

Without this change, the unit tests are failing because the Docker
container has been updated to 1.2.0, and PHPUnit is configured to
treat notices as test failures.

See: https://github.com/nats-io/gnatsd/commit/17fecd4c9be6592de42d1ff5c8c2659a5d0382a5#diff-91bbeda7eb98a7adc57b9e47e2cf5c2b

Commit:
adds checks for optional fields in server response

Commit:
adds restore_error_handler before throwing exception

Commit:
corrected restore_error_handler call

tomponline commented:
@repejota this has been updated to work with unit tests on latest gnatsd server now, are you OK to merge?

Commit:
Revert "corrected restore_error_handler call"

This reverts commit c604bd3550fa523ff5ada078bae2a883c959c061.

Commit:
Revert "adds restore_error_handler before throwing exception"

This reverts commit b77907aff90b7810cea48fff1b4e47124224d8e7.

Commit:
Revert "adds checks for optional fields in server response"

This reverts commit b1f925a9fa2ce0c3c7b78dba8f09c0072c06aab5.

Commit:
Adds restore_error_handler to isTLSRequired error handler

tomponline commented:
Hi @repejota have you had a chance to look at this since you've come back from holiday? Would be great to get this merged if you're happy with it. Trying to avoid having to fork. Cheers

repejota reviewed (approved):
Yay!
Really good work! I have nothing to comment, this is just a great pull request :)

tomponline commented:
Awesome thanks @repejota

repejota commented:
Gonna be adding a few things (mostly documentation and help files for
github) and will release 0.9 version.

Again thanks for your work mate! :D

On Wed, Sep 5, 2018 at 2:56 PM TomP <notifications@github.com> wrote:

> Awesome thanks @repejota <https://github.com/repejota>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/repejota/phpnats/pull/122#issuecomment-418720392>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACPQa4gmMnvi0GGPVZ-tF9OkT_UuwRAks5uX8oPgaJpZM4VZlKJ>
> .
>


-- 
--
Raül Pérez, repejota@gmail.com

onyxg commented:
@tomponline this is gold... will implement with our Golang stack and advise

mardachevdv commented:
Hi @repejota 
Could you release new version with this changes?

sazo commented:
@repejota Really need this to go live. We are a company using the package but we cant go to production before TLS is supported. :)
